pkg/controller/util/node/controller_utils.go

```diff
@@ -42,16 +42,15 @@ import (
 // DeletePods will delete all pods from master running on given node,
 // and return true if any pods were deleted, or were found pending
 // deletion.
-func DeletePods(kubeClient clientset.Interface, pods []v1.Pod, recorder record.EventRecorder, nodeName, nodeUID string, daemonStore appsv1listers.DaemonSetLister) (bool, error) {
+func DeletePods(kubeClient clientset.Interface, pods []*v1.Pod, recorder record.EventRecorder, nodeName, nodeUID string, daemonStore appsv1listers.DaemonSetLister) (bool, error) {
 	remaining := false
 	var updateErrList []error
 
 	if len(pods) > 0 {
 		RecordNodeEvent(recorder, nodeName, nodeUID, v1.EventTypeNormal, "DeletingAllPods", fmt.Sprintf("Deleting all Pods from Node %v.", nodeName))
 	}
 
-	for i := range pods {
-		pod := &pods[i]
+	for _, pod := range pods {
```

Comment:
It seems this change isn't correct. Here below, you call SetPodTerminationReason,
which is modifying the pod.

So it works now, when what is passed to this method has just been explicitly listed from server. But it will no longer work when listing will be done from informer what you're planning to do.

Basically, what you need to do, is to DeepCopy the pod before you start modifying it.

You need to do a similar audit in all other places.

Reply:
I've added DeepCopy to `DeletePods` and `MarkPodsNotReady`.